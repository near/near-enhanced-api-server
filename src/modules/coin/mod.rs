use actix_web::{HttpResponse, error};
use actix_web_validator::{Error, PathConfig};
use paperclip::actix::web;

mod data_provider;
mod resources;
mod schemas;

#[derive(serde::Serialize)]
pub struct ValidationErrorJsonPayload {
    pub message: String,
    pub fields: Vec<String>,
}

/// Custom error handler
impl From<&validator::ValidationErrors> for ValidationErrorJsonPayload {
    fn from(error: &validator::ValidationErrors) -> Self {
        ValidationErrorJsonPayload {
            message: "Validation error".to_owned(),
            fields: error.field_errors().iter().map(|(field, _)| field.to_string()).collect(),
        }
    }
}

pub(crate) fn register_services(app: &mut web::ServiceConfig) {
    app
        .app_data(PathConfig::default().error_handler(|err, _| {
            let json_error = match &err {
                Error::Validate(error) => ValidationErrorJsonPayload::from(error),
                _ => ValidationErrorJsonPayload { message: err.to_string(), fields: Vec::new() },
            };
            error::InternalError::from_response(err, HttpResponse::Conflict().json(json_error)).into()
        }))
        .service(
        web::resource("/accounts/{account_id}/coins/NEAR")
            .route(web::get().to(resources::get_near_balance)),
    )
    .service(
        web::resource("/accounts/{account_id}/coins")
            .route(web::get().to(resources::get_coin_balances)),
    )
    .service(
        web::resource("/accounts/{account_id}/coins/{contract_account_id}")
            .route(web::get().to(resources::get_coin_balances_by_contract)),
    )
    .service(
        web::resource("/accounts/{account_id}/coins/NEAR/history")
            .route(web::get().to(resources::get_near_history)),
    )
    .service(
        web::resource("/accounts/{account_id}/coins/{contract_account_id}/history")
            .route(web::get().to(resources::get_coin_history)),
    )
    .service(
        web::resource("/nep141/metadata/{contract_account_id}")
            .route(web::get().to(resources::get_ft_contract_metadata)),
    );
}
