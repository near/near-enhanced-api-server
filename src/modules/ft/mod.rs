use paperclip::actix::web;

mod data_provider;
mod resources;
mod schemas;

pub(crate) fn register_services(app: &mut web::ServiceConfig) {
    // I'll drop these comments, leaving it only because it's easier for the review
    // app.service(
    //     web::resource("/accounts/{account_id}/NEAR")
    //         .route(web::get().to(resources::get_near_balance)),
    // )
    app.service(
        web::resource("/accounts/{account_id}/FT").route(web::get().to(resources::get_ft_balances)),
    )
    .service(
        web::resource("/accounts/{account_id}/FT/{contract_account_id}")
            .route(web::get().to(resources::get_ft_balance_by_contract)),
    )
    // .service(
    //     web::resource("/accounts/{account_id}/NEAR/history")
    //         .route(web::get().to(resources::get_near_history)),
    // )
    .service(
        web::resource("/accounts/{account_id}/FT/{contract_account_id}/history")
            .route(web::get().to(resources::get_ft_history)),
    )
    .service(
        web::resource("/nep141/metadata/{contract_account_id}")
            .route(web::get().to(resources::get_ft_metadata)),
    );
}
