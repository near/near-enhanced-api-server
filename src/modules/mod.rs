use crate::{db_helpers, errors, types};

pub(crate) mod coin;
pub(crate) mod nft;

pub(crate) async fn check_account_exists(
    rpc_client: &near_jsonrpc_client::JsonRpcClient,
    account_id: &near_primitives::types::AccountId,
    block_height: u64,
) -> crate::Result<()> {
    let request = near_jsonrpc_client::methods::query::RpcQueryRequest {
        block_reference: near_primitives::types::BlockId::Height(block_height).into(),
        request: near_primitives::views::QueryRequest::ViewAccount {
            account_id: account_id.clone(),
        },
    };
    for _ in 0..5 {
        match rpc_client.call(&request).await {
            Err(near_jsonrpc_client::errors::JsonRpcError::ServerError(
                near_jsonrpc_client::errors::JsonRpcServerError::HandlerError(
                    near_jsonrpc_client::methods::query::RpcQueryError::UnknownAccount { .. },
                ),
            )) => {
                return Err(errors::ErrorKind::InvalidInput(format!(
                    "account_id {} does not exist at block_height {}",
                    account_id, block_height
                ))
                .into())
            }
            Err(err) => {
                tracing::warn!(target: crate::LOGGER_MSG, "Checking account existence via JSON RPC failed with: {:?}. Re-trying immediatelly", err);
                continue;
            }
            Ok(response) => {
                if let near_jsonrpc_primitives::types::query::QueryResponseKind::ViewAccount(_) =
                    response.kind
                {
                    return Ok(());
                } else {
                    tracing::warn!(target: crate::LOGGER_MSG, "Checking account existence returned invalid response: {:?}. Re-trying immediatelly", response);
                    continue;
                }
            }
        }
    }
    Err(errors::ErrorKind::InternalError(format!(
        "could not check if account_id {} exists after several attemps",
        account_id
    ))
    .into())
}

pub(crate) async fn check_and_get_history_pagination_params(
    pool: &sqlx::Pool<sqlx::Postgres>,
    pagination_params: types::query_params::HistoryPaginationParams,
) -> crate::Result<types::query_params::HistoryPagination> {
    types::query_params::check_limit(pagination_params.limit)?;
    let pagination = types::query_params::Pagination::from(pagination_params);
    // if pagination_params.after_block_height.is_some() && pagination_params.after_timestamp_nanos.is_some() {
    //     return Err(errors::ErrorKind::InvalidInput(
    //         "Both block_height and block_timestamp_nanos found. Please provide only one of values"
    //             .to_string(),
    //     )
    //         .into());
    // }
    // TODO PHASE 2 take the block from pagination_params
    let block = db_helpers::get_last_block(pool).await?;
    Ok(types::query_params::HistoryPagination {
        block_height: block.height,
        block_timestamp: block.timestamp,
        limit: pagination.limit,
    })
}

#[cfg(test)]
mod tests {
    use crate::db_helpers;

    pub(crate) async fn init_db() -> sqlx::Pool<sqlx::Postgres> {
        dotenv::dotenv().ok();
        let db_url = &std::env::var("DATABASE_URL").expect("failed to get database url");

        sqlx::PgPool::connect(db_url)
            .await
            .expect("failed to connect to the database")
    }

    pub(crate) fn init_rpc() -> near_jsonrpc_client::JsonRpcClient {
        dotenv::dotenv().ok();
        let rpc_url = &std::env::var("RPC_URL").expect("failed to get RPC url");
        let connector = near_jsonrpc_client::JsonRpcClient::new_client();
        connector.connect(rpc_url)
    }

    pub(crate) fn get_block() -> db_helpers::Block {
        db_helpers::Block {
            timestamp: 1655571176644255779,
            height: 68000000,
        }
    }
}
