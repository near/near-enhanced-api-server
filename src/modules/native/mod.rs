use paperclip::actix::web;

mod data_provider;
mod resources;
mod schemas;

pub(crate) fn register_services(app: &mut web::ServiceConfig) {
    app.service(
        web::resource("/accounts/{account_id}/balances/NEAR")
            .route(web::get().to(resources::get_near_balance)),
    )
    .service(
        web::resource("/accounts/{account_id}/balances/NEAR/history")
            .route(web::get().to(resources::get_near_history)),
    );
}
