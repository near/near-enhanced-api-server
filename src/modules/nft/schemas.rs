use paperclip::actix::Apiv2Schema;

use crate::types;

// *** Requests ***

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct NftCountsRequest {
    pub account_id: types::AccountId,
}

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct NftCollectionRequest {
    pub account_id: types::AccountId,
    pub contract_account_id: types::AccountId,
}

// duplicate in each folder
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct MetadataRequest {
    pub contract_account_id: types::AccountId,
}

/// `token_id` is available at `NftCollectionByContractResponse`
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct NftRequest {
    pub contract_account_id: types::AccountId,
    pub token_id: String,
}

// *** Responses ***

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct NftCountsResponse {
    pub nft_counts: Vec<NftCount>,
    pub block_timestamp_nanos: types::U64,
    pub block_height: types::U64,
}

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct NftsResponse {
    pub nfts: Vec<Nft>,
    pub contract_metadata: NftContractMetadata,
    pub block_timestamp_nanos: types::U64,
    pub block_height: types::U64,
}

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct NftResponse {
    pub nft: Nft,
    pub contract_metadata: NftContractMetadata,
    pub block_timestamp_nanos: types::U64,
    pub block_height: types::U64,
}

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct HistoryResponse {
    pub history: Vec<HistoryItem>,
    pub nft: Nft,
    pub block_timestamp_nanos: types::U64,
    pub block_height: types::U64,
}

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct MetadataResponse {
    pub metadata: NftContractMetadata,
    pub block_timestamp_nanos: types::U64,
    pub block_height: types::U64,
}

// ---

/// This type describes the history of NFT movements.
/// Note, it's not attached to any user, it's the whole history of NFT movements.
/// `cause` is one of ["mint", "transfer", "burn"]
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct HistoryItem {
    pub cause: String,
    pub old_account_id: Option<types::AccountId>,
    pub new_account_id: Option<types::AccountId>,
    // TODO PHASE 2 add index here
    // pub index: types::U128,
    pub status: String,
    pub block_timestamp_nanos: types::U64,
    pub block_height: types::U64,
}

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct NftCount {
    pub contract_account_id: types::AccountId,
    pub nft_count: u32,
    // TODO PHASE 1 naming.
    pub last_updated_at_timestamp_nanos: types::U128,
    pub contract_metadata: NftContractMetadata,
}

/// The type for Non Fungible Token Contract Metadata. Inspired by
/// https://nomicon.io/Standards/Tokens/NonFungibleToken/Metadata
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct NftContractMetadata {
    pub spec: String,              // required, essentially a version like "nft-1.0.0"
    pub name: String,              // required, ex. "Mosaics"
    pub symbol: String,            // required, ex. "MOSIAC"
    pub icon: Option<String>,      // Data URL
    pub base_uri: Option<String>, // Centralized gateway known to have reliable access to decentralized data_provider assets referenced by `reference` or `media` URLs
    pub reference: Option<String>, // URL to a JSON file with more info
    pub reference_hash: Option<String>, // Base64-encoded sha256 hash of JSON from reference field. Required if `reference` is included.
}

/// The type for Non Fungible Token. Inspired by
/// https://nomicon.io/Standards/Tokens/NonFungibleToken/Metadata
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct Nft {
    pub token_id: String,
    pub owner_account_id: String,
    pub metadata: NftMetadata,
    // TODO PHASE 1 do we want to show them? People often put here weird things
    // pub approved_account_ids: Option<std::collections::HashMap<AccountId, u64>>,
}

/// The type for Non Fungible Token Metadata. Inspired by
/// https://nomicon.io/Standards/Tokens/NonFungibleToken/Metadata
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct NftMetadata {
    pub title: Option<String>, // ex. "Arch Nemesis: Mail Carrier" or "Parcel #5055"
    pub description: Option<String>, // free-form description
    pub media: Option<String>, // URL to associated media, preferably to decentralized, content-addressed data_provider
    pub media_hash: Option<String>, // Base64-encoded sha256 hash of content referenced by the `media` field. Required if `media` is included.
    pub copies: Option<u64>, // number of copies of this set of metadata in existence when token was minted.
    pub issued_at: Option<String>, // ISO 8601 datetime when token was issued or minted
    pub expires_at: Option<String>, // ISO 8601 datetime when token expires
    pub starts_at: Option<String>, // ISO 8601 datetime when token starts being valid
    pub updated_at: Option<String>, // ISO 8601 datetime when token was last updated
    pub extra: Option<String>, // anything extra the NFT wants to data_provider on-chain. Can be stringified JSON.
    pub reference: Option<String>, // URL to an off-chain JSON file with more info.
    pub reference_hash: Option<String>, // Base64-encoded sha256 hash of JSON from reference field. Required if `reference` is included.
}
