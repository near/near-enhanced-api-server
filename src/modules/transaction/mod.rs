use paperclip::actix::web;

// mod data_provider;
mod resources;
mod schemas;

pub(crate) fn register_services(app: &mut web::ServiceConfig) {
    app.service(
        web::resource("/transaction/{transaction_hash}")
            .route(web::get().to(resources::get_transaction_by_tx_hash)),
    )
    .service(
        web::resource("/transaction/receipt/{receipt_id}")
            .route(web::get().to(resources::get_transaction_by_receipt)),
    )
    .service(
        web::resource("/transactions/{account_id}")
            .route(web::get().to(resources::get_transactions_by_account)),
    )
    .service(
        web::resource("/transactions/{account_id}/{contract_id}")
            .route(web::get().to(resources::get_transactions_by_account_on_contract)),
    )
    .service(
        web::resource("/receipts/{transaction_hash}")
            .route(web::get().to(resources::get_receipts_by_tx_hash)),
    )
    .service(
        web::resource("/receipts/actions/{account_id}")
            .route(web::get().to(resources::get_action_receipts_by_account)),
    )
    .service(
        web::resource("/receipts/actions/{account_id}/{contract_id}")
            .route(web::get().to(resources::get_action_receipts_by_account_on_contract)),
    );
}
