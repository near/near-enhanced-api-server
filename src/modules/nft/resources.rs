use paperclip::actix::{
    api_v2_operation,
    web::{self, Json},
};

use crate::{db_helpers, modules, types};

use super::schemas;

#[api_v2_operation(tags(NFT))]
/// Get user's NFT collection overview
///
/// For the given `account_id`, this endpoint returns
/// the number of NFTs grouped by `contract_account_id`, together with the corresponding NFT contract metadata.
/// The NFT contract will be present in the response if the `account_id` has at least one NFT there.
///
/// **Limitations**
/// * We currently provide the most recent 100 items.
///   Full-featured pagination will be provided soon.
pub async fn get_nft_collection_overview(
    pool_explorer: web::Data<db_helpers::ExplorerPool>,
    rpc_client: web::Data<near_jsonrpc_client::JsonRpcClient>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::NftCountsRequest>,
    limit_params: web::Query<types::query_params::LimitParams>,
) -> crate::Result<Json<schemas::NftCountsResponse>> {
    let limit = types::query_params::checked_get_limit(limit_params.limit)?;
    let block = db_helpers::get_last_block(&pool_explorer).await?;
    modules::check_account_exists(&rpc_client, &request.account_id.0, block.height).await?;

    Ok(Json(schemas::NftCountsResponse {
        // TODO We can save metadata in the DB and update once in 10 minutes
        nft_counts: super::data_provider::get_nfts_count(
            &pool_explorer,
            &rpc_client,
            &block,
            &request.account_id.0,
            limit,
        )
        .await?,
        block_timestamp_nanos: types::U64::from(block.timestamp),
        block_height: types::U64::from(block.height),
    }))
}

#[api_v2_operation(tags(NFT))]
/// Get user's NFT collection by contract
///
/// This endpoint returns the list of NFTs with full details for the given `account_id`, NFT `contract_account_id`.
/// You can use the `token_id` from this response and then request the NFT history for that token.
///
/// **Limitations**
/// * We currently provide the most recent 100 items.
///   Full-featured pagination will be provided soon.
pub async fn get_nft_collection_by_contract(
    pool_explorer: web::Data<db_helpers::ExplorerPool>,
    rpc_client: web::Data<near_jsonrpc_client::JsonRpcClient>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::NftCollectionRequest>,
    limit_params: web::Query<types::query_params::LimitParams>,
) -> crate::Result<Json<schemas::NftsResponse>> {
    let limit = types::query_params::checked_get_limit(limit_params.limit)?;
    let block = db_helpers::get_last_block(&pool_explorer).await?;
    modules::check_account_exists(&rpc_client, &request.account_id.0, block.height).await?;

    Ok(Json(schemas::NftsResponse {
        nfts: super::data_provider::get_nfts_by_contract(
            &rpc_client,
            request.contract_account_id.0.clone(),
            request.account_id.0.clone(),
            block.height,
            limit,
        )
        .await?,
        contract_metadata: super::data_provider::get_nft_contract_metadata(
            &rpc_client,
            request.contract_account_id.0.clone(),
            block.height,
        )
        .await?,
        block_timestamp_nanos: types::U64::from(block.timestamp),
        block_height: types::U64::from(block.height),
    }))
}

#[api_v2_operation(tags(NFT))]
/// Get NFT
///
/// This endpoint returns detailed information on the NFT
/// for the given `token_id`, NFT `contract_account_id`, `block_timestamp_nanos`/`block_height`.
pub async fn get_nft(
    pool_explorer: web::Data<db_helpers::ExplorerPool>,
    rpc_client: web::Data<near_jsonrpc_client::JsonRpcClient>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::NftRequest>,
    block_params: web::Query<types::query_params::BlockParams>,
) -> crate::Result<Json<schemas::NftResponse>> {
    let block = db_helpers::checked_get_block(&pool_explorer, &block_params).await?;

    Ok(Json(schemas::NftResponse {
        nft: super::data_provider::get_nft(
            &rpc_client,
            request.contract_account_id.0.clone(),
            request.token_id.clone(),
            block.height,
        )
        .await?,
        contract_metadata: super::data_provider::get_nft_contract_metadata(
            &rpc_client,
            request.contract_account_id.0.clone(),
            block.height,
        )
        .await?,
        block_timestamp_nanos: types::U64::from(block.timestamp),
        block_height: types::U64::from(block.height),
    }))
}

#[api_v2_operation(tags(NFT))]
/// Get NFT history
///
/// This endpoint returns the transaction history for the given NFT.
/// **Note:** The result is centered around the history of the specific NFT and will return list of its passing owners.
///
/// **Limitations**
/// * For now, we only support NFT contracts that implement the Events NEP standard.
/// * We currently provide the most recent 100 items.
///   Full-featured pagination will be provided soon.
pub async fn get_nft_history(
    pool_explorer: web::Data<db_helpers::ExplorerPool>,
    rpc_client: web::Data<near_jsonrpc_client::JsonRpcClient>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::NftRequest>,
    limit_params: web::Query<types::query_params::LimitParams>,
) -> crate::Result<Json<schemas::NftHistoryResponse>> {
    let limit = types::query_params::checked_get_limit(limit_params.limit)?;
    let block = db_helpers::get_last_block(&pool_explorer).await?;

    Ok(Json(schemas::NftHistoryResponse {
        history: super::data_provider::get_nft_history(
            &pool_explorer,
            &request.contract_account_id.0,
            &request.token_id,
            &block,
            limit,
        )
        .await?,
        nft: super::data_provider::get_nft(
            &rpc_client,
            request.contract_account_id.0.clone(),
            request.token_id.clone(),
            block.height,
        )
        .await?,
        block_timestamp_nanos: types::U64::from(block.timestamp),
        block_height: types::U64::from(block.height),
    }))
}

#[api_v2_operation(tags(NFT))]
/// Get NFT contract metadata
///
/// This endpoint returns the metadata for a given NFT contract and `block_timestamp_nanos`/`block_height`.
/// **Note:** This is contract-wide metadata. Each NFT also has its own metadata.
pub async fn get_nft_contract_metadata(
    pool_explorer: web::Data<db_helpers::ExplorerPool>,
    rpc_client: web::Data<near_jsonrpc_client::JsonRpcClient>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::MetadataRequest>,
    block_params: web::Query<types::query_params::BlockParams>,
) -> crate::Result<Json<schemas::MetadataResponse>> {
    let block = db_helpers::checked_get_block(&pool_explorer, &block_params).await?;

    Ok(Json(schemas::MetadataResponse {
        metadata: super::data_provider::get_nft_contract_metadata(
            &rpc_client,
            request.contract_account_id.0.clone(),
            block.height,
        )
        .await?,
        block_timestamp_nanos: types::U64::from(block.timestamp),
        block_height: types::U64::from(block.height),
    }))
}
