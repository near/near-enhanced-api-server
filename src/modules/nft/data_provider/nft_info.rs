use std::str::FromStr;

use crate::modules::nft;
use crate::{db_helpers, errors, rpc_helpers, types};
use serde::{Deserialize, Serialize};

// TODO PHASE 2 pagination by artificial index added to assets__non_fungible_token_events
pub(crate) async fn get_nfts_count(
    db_helpers::ExplorerPool(pool_explorer): &db_helpers::ExplorerPool,
    rpc_client: &near_jsonrpc_client::JsonRpcClient,
    block: &db_helpers::Block,
    account_id: &near_primitives::types::AccountId,
    limit: u32,
) -> crate::Result<Vec<nft::schemas::NftCount>> {
    let query = r"
        WITH relevant_events AS (
            SELECT emitted_at_block_timestamp, token_id, emitted_by_contract_account_id, token_old_owner_account_id, token_new_owner_account_id
            FROM assets__non_fungible_token_events
                JOIN execution_outcomes ON assets__non_fungible_token_events.emitted_for_receipt_id = execution_outcomes.receipt_id
            WHERE
                -- if it works slow, we need to create table daily_nft_count_by_contract_and_user, and this query will run only over the last day
                -- emitted_at_block_timestamp > start_of_day AND
                emitted_at_block_timestamp <= $2::numeric(20, 0)
                AND execution_outcomes.status IN ('SUCCESS_VALUE', 'SUCCESS_RECEIPT_ID')
                AND (token_new_owner_account_id = $1 OR token_old_owner_account_id = $1)
        ),
        outgoing_events_count AS (
            SELECT emitted_by_contract_account_id, count(*) * -1 cnt FROM relevant_events
            WHERE token_old_owner_account_id = $1
            GROUP BY emitted_by_contract_account_id
        ),
        ingoing_events_count AS (
            SELECT emitted_by_contract_account_id, count(*) cnt FROM relevant_events
            WHERE token_new_owner_account_id = $1
            GROUP BY emitted_by_contract_account_id
        ),
        counts AS (
            SELECT ingoing_events_count.emitted_by_contract_account_id,
                -- coalesce changes null to the given parameter
                coalesce(ingoing_events_count.cnt, 0) + coalesce(outgoing_events_count.cnt, 0) cnt
            FROM ingoing_events_count FULL JOIN outgoing_events_count
                ON ingoing_events_count.emitted_by_contract_account_id = outgoing_events_count.emitted_by_contract_account_id
        ),
        counts_with_timestamp AS (
            SELECT distinct ON (counts.emitted_by_contract_account_id) counts.emitted_by_contract_account_id contract_id,
                cnt count,
                emitted_at_block_timestamp last_updated_at_timestamp
            FROM counts JOIN relevant_events ON counts.emitted_by_contract_account_id = relevant_events.emitted_by_contract_account_id
            WHERE cnt > 0
            ORDER BY counts.emitted_by_contract_account_id, emitted_at_block_timestamp DESC
        )
        SELECT * FROM counts_with_timestamp
        -- WHERE last_updated_at_timestamp < $3::numeric(20, 0) -- phase 2 pagination will be covered here
        ORDER BY last_updated_at_timestamp DESC
        LIMIT $3::numeric(20, 0)
    ";

    let info_by_contract = db_helpers::select_retry_or_panic::<super::models::NftCount>(
        pool_explorer,
        query,
        &[
            account_id.to_string(),
            block.timestamp.to_string(),
            limit.to_string(),
        ],
    )
    .await?;

    let mut result: Vec<nft::schemas::NftCount> = vec![];
    for info in info_by_contract {
        if let Ok(contract_id) = near_primitives::types::AccountId::from_str(&info.contract_id) {
            let metadata = super::metadata::get_nft_contract_metadata(
                rpc_client,
                contract_id.clone(),
                block.height,
            )
            .await
            .unwrap_or_else(|_| super::metadata::get_default_nft_contract_metadata());
            result.push(nft::schemas::NftCount {
                contract_account_id: contract_id.into(),
                nft_count: info.count as u32,
                last_updated_at_timestamp_nanos: types::numeric::to_u128(
                    &info.last_updated_at_timestamp,
                )?
                .into(),
                contract_metadata: metadata,
            });
        }
    }
    Ok(result)
}

pub(crate) async fn get_nfts_by_contract(
    rpc_client: &near_jsonrpc_client::JsonRpcClient,
    contract_id: near_primitives::types::AccountId,
    account_id: near_primitives::types::AccountId,
    block_height: u64,
    limit: u32,
) -> crate::Result<Vec<nft::schemas::Nft>> {
    // TODO PHASE 2 pagination
    // RPC supports pagination, but the order is defined by the each contract and we can't control it.
    // For now, we are ready to serve only the first page
    // Later, I feel we need to load NFT (each token) metadata to the DB,
    // right after that we can stop using RPC here.
    // Or, maybe we want to delegate this task fully to the contracts?
    let request = rpc_helpers::get_function_call_request(
        block_height,
        contract_id.clone(),
        "nft_tokens_for_owner",
        // https://nomicon.io/Standards/Tokens/NonFungibleToken/Enumeration
        serde_json::json!({ "account_id": account_id, "from_index": "0", "limit": limit }),
    );
    let response =
        rpc_helpers::wrapped_call(rpc_client, request, block_height, &contract_id).await?;

    let tokens = serde_json::from_slice::<Vec<Token>>(&response.result)?;
    let mut result = vec![];
    for token in tokens {
        result.push(nft::schemas::Nft::try_from(token)?);
    }
    Ok(result)
}

pub(crate) async fn get_nft(
    rpc_client: &near_jsonrpc_client::JsonRpcClient,
    contract_id: near_primitives::types::AccountId,
    token_id: String,
    block_height: u64,
) -> crate::Result<nft::schemas::Nft> {
    let request = rpc_helpers::get_function_call_request(
        block_height,
        contract_id.clone(),
        "nft_token",
        serde_json::json!({ "token_id": token_id }),
    );
    let response =
        rpc_helpers::wrapped_call(rpc_client, request, block_height, &contract_id).await?;

    match serde_json::from_slice::<Option<Token>>(&response.result)? {
        None => Err(errors::ErrorKind::InvalidInput(format!(
            "Token `{}` does not exist in contract `{}`, block_height {}",
            token_id, contract_id, block_height
        ))
        .into()),
        Some(token) => nft::schemas::Nft::try_from(token),
    }
}

// Taken from https://github.com/near/near-sdk-rs/blob/master/near-contract-standards/src/non_fungible_token/token.rs
/// Note that token IDs for NFTs are strings on NEAR. It's still fine to use autoincrementing numbers as unique IDs if desired, but they should be stringified. This is to make IDs more future-proof as chain-agnostic conventions and standards arise, and allows for more flexibility with considerations like bridging NFTs across chains, etc.
pub type TokenId = String;

/// In this implementation, the Token struct takes two extensions standards (metadata and approval) as optional fields, as they are frequently used in modern NFTs.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct Token {
    pub token_id: TokenId,
    pub owner_id: types::AccountId,
    pub metadata: Option<nft::schemas::NftMetadata>,
    pub approved_account_ids: Option<std::collections::HashMap<types::AccountId, u64>>,
}

impl TryFrom<Token> for nft::schemas::Nft {
    type Error = errors::Error;

    fn try_from(token: Token) -> crate::Result<Self> {
        let metadata = token.metadata.ok_or_else(|| {
            errors::ErrorKind::ContractError(
                "The contract did not provide NFT Metadata which is a required part of NFT NEP 171"
                    .to_string(),
            )
        })?;

        Ok(Self {
            token_id: token.token_id,
            owner_account_id: token.owner_id.0.to_string(),
            metadata,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::modules::tests::*;

    #[tokio::test]
    async fn test_nfts_count() {
        let pool_explorer = init_explorer_db().await;
        let rpc_client = init_rpc();
        let block = get_block();
        let account = near_primitives::types::AccountId::from_str("blondjesus.near").unwrap();

        let nft_count = get_nfts_count(&pool_explorer, &rpc_client, &block, &account, 10).await;
        insta::assert_debug_snapshot!(nft_count);
    }

    #[tokio::test]
    async fn test_nfts_count_empty() {
        let pool_explorer = init_explorer_db().await;
        let rpc_client = init_rpc();
        let block = get_block();
        let account = near_primitives::types::AccountId::from_str("cucumber.near").unwrap();

        let nft_count = get_nfts_count(&pool_explorer, &rpc_client, &block, &account, 10)
            .await
            .unwrap();
        assert!(nft_count.is_empty());
    }

    #[tokio::test]
    async fn test_nfts_count_with_contracts_with_no_metadata() {
        let pool_explorer = init_explorer_db().await;
        let rpc_client = init_rpc();
        let block = db_helpers::Block {
            timestamp: 1655571176644255779,
            height: 68000000,
        };
        let account = near_primitives::types::AccountId::from_str("vlad.near").unwrap();

        let nft_count = get_nfts_count(&pool_explorer, &rpc_client, &block, &account, 10).await;
        insta::assert_debug_snapshot!(nft_count);
    }

    #[tokio::test]
    async fn test_nfts_count_with_no_failed_receipts_in_result() {
        let pool_explorer = init_explorer_db().await;
        let rpc_client = init_rpc();
        let block = get_block();
        let account = near_primitives::types::AccountId::from_str("kbneoburner3.near").unwrap();

        let nft_count = get_nfts_count(&pool_explorer, &rpc_client, &block, &account, 10).await;
        insta::assert_debug_snapshot!(nft_count);
    }

    #[tokio::test]
    async fn test_nfts_by_contract() {
        let rpc_client = init_rpc();
        let block = get_block();
        let contract =
            near_primitives::types::AccountId::from_str("billionairebullsclub.near").unwrap();
        let account = near_primitives::types::AccountId::from_str("olenavorobei.near").unwrap();

        let nfts = get_nfts_by_contract(&rpc_client, contract, account, block.height, 4).await;
        insta::assert_debug_snapshot!(nfts);
    }

    #[tokio::test]
    async fn test_nft() {
        let rpc_client = init_rpc();
        let block = get_block();
        let contract = near_primitives::types::AccountId::from_str("x.paras.near").unwrap();
        let token = "415815:1".to_string();

        let nft = get_nft(&rpc_client, contract, token, block.height).await;
        insta::assert_debug_snapshot!(nft);
    }

    #[tokio::test]
    async fn test_nft_does_not_exist() {
        let rpc_client = init_rpc();
        let block = get_block();
        let contract = near_primitives::types::AccountId::from_str("x.paras.near").unwrap();
        let token = "no_such_token".to_string();

        let nft = get_nft(&rpc_client, contract, token, block.height).await;
        insta::assert_debug_snapshot!(nft);
    }
}
