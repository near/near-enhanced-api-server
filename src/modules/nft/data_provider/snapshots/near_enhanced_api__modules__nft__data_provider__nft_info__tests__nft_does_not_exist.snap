---
source: src/modules/nft/data_provider/nft_info.rs
expression: nft
---
Err(
    Error {
        code: 400,
        message: "Invalid Input: Token `no_such_token` does not exist in contract `x.paras.near`, block_height 80500000",
        retriable: false,
    },
)
